ELFSage/test.lean

```diff
@@ -0,0 +1,121 @@
+import Lean
+open Lean Parser Command
+
+declare_syntax_cat enum_case
+syntax (docComment)? " | " ident " => " term : enum_case
+
+private def split (case : TSyntax `enum_case) : Ident × Term × (Lean.TSyntax `Lean.Parser.Command.docComment) :=
+  match case with
+    | `(enum_case| | $name => $value) => (name, value, emptyComment)
+    | `(enum_case| $comment:docComment | $name => $value) => (name, value, comment)
+    | _ => unreachable!
+  where commentLeader : Lean.Syntax := Lean.Syntax.atom SourceInfo.none "/--"
+        commentBody : Lean.Syntax := Lean.Syntax.atom SourceInfo.none "undocumented -/"
+        emptyComment : Lean.TSyntax `Lean.Parser.Command.docComment := Lean.TSyntax.mk $
+          Lean.Syntax.node SourceInfo.none `Lean.Parser.Command.docComment #[commentLeader, commentBody]
+
+private def splitArray (cases : TSyntaxArray `enum_case) : (Array Ident) × (Array Term) × (Array (Lean.TSyntax `Lean.Parser.Command.docComment)) :=
+  cases.foldl (init := (#[], #[], #[])) fun (names, values, comments) => fun (case) =>
+    let (name, value, comment) := split case
+    (names.push name, values.push value, comments.push comment)
+
+/--
+The `enum` macro creates named constants for each case of an enumeration type.
+A default case is created for any value not explicitly enumerated.
+
+For now, the enum value type must be `Nat`. This may be generalized to support
+`Fin n` in the future.
+-/
+macro "enum" e:ident "(" n:ident " : " ty:term ")" (" where")? cases:enum_case* : command => do
+  let eIdent (suffix : Name) := mkIdentFrom e (e.getId ++ suffix)
+  let toStrLit (n : Name) := Lean.Syntax.mkStrLit n.toString
+
+  let defaultCase := cases.back
+  let enumCases := cases.pop
+  let (names, values, comments) := splitArray enumCases
+  let (defaultName, defaultValue, defaultComment) := split defaultCase
+  let valuesList: TSyntax `term := Lean.quote values.toList
+  let rawNames := names.map (·.getId)
+  let strNames := rawNames.map toStrLit
+  let prefixedNames := rawNames.map (eIdent ·)
+
+  if ¬ n == defaultValue then
+    -- TODO is there a better way to error? Panicking results in a terrible authoring experience.
```

Comment:
I think Macro.throwErrorAt might do the trick: https://leanprover-community.github.io/mathlib4_docs/Init/Prelude.html#Lean.Macro.throwErrorAt